cwltool/singularity.py

```diff
@@ -79,39 +86,55 @@ def get_image(dockerRequirement,  # type: Dict[Text, Text]
 
         if "dockerImageId" not in dockerRequirement and "dockerPull" in dockerRequirement:
             match = re.search(pattern=r'([a-z]*://)', string=dockerRequirement["dockerPull"])
-            candidate = _normalize_image_id(dockerRequirement['dockerPull'])
-            candidates.append(candidate)
-            dockerRequirement['dockerImageId'] = candidate
+            candidate_image = _normalize_image_id(dockerRequirement['dockerPull'])
+            candidates.append(candidate_image)
+            candidate_sif = _normalize_sif_id(dockerRequirement['dockerPull'])
+            candidates.append(candidate_sif)
+            dockerRequirement['dockerImageId'] = candidate_image
             if not match:
                 dockerRequirement["dockerPull"] = "docker://" + dockerRequirement["dockerPull"]
         elif "dockerImageId" in dockerRequirement:
             candidates.append(dockerRequirement['dockerImageId'])
             candidates.append(_normalize_image_id(dockerRequirement['dockerImageId']))
 
         # check if Singularity image is available in $SINGULARITY_CACHEDIR
+        # or any subdirs created in cachedir
         targets = [os.getcwd()]
         for env in ("SINGULARITY_CACHEDIR", "SINGULARITY_PULLFOLDER"):
             if env in os.environ:
                 targets.append(os.environ[env])
         for target in targets:
-            for candidate in candidates:
-                path = os.path.join(target, candidate)
-                if os.path.isfile(path):
-                    _logger.info(
-                        "Using local copy of Singularity image found in %s",
-                        target)
-                    dockerRequirement["dockerImageId"] = path
-                    found = True
-
+            for dirpath, subdirs, files in os.walk(target):
+                for sif in files:
+                    if sif in candidates:
+                        path = os.path.join(dirpath, sif)
+                        if os.path.isfile(path):
+                            _logger.info(
+                                "Using local copy of Singularity image found in %s",
+                                dirpath)
+                            dockerRequirement["dockerImageId"] = path
+                            found = True
         if (force_pull or not found) and pull_image:
             cmd = []  # type: List[Text]
             if "dockerPull" in dockerRequirement:
-                cmd = ["singularity", "pull", "--force", "--name",
-                       str(dockerRequirement["dockerImageId"]),
-                       str(dockerRequirement["dockerPull"])]
-                _logger.info(Text(cmd))
-                check_call(cmd, stdout=sys.stderr)
-                found = True
+                if "SINGULARITY_PULLFOLDER" in os.environ:
+                    pull_folder = str(os.environ["SINGULARITY_PULLFOLDER"])
+                    cmd = ["singularity", "pull", "--force", "--name",
```

Comment:
if image is not found, and $SINGULARITY_PULLFOLDER is in environment, write the new image to this folder. Else write it locally.